File: src/algorithms/hash-maps/counting/majority-element/sources/majority-element.py
# Majority Element — find the element that appears more than n/2 times using a frequency map
def majority_element(numbers):
    frequency_map = {}  # @step:initialize
    threshold = len(numbers) // 2  # @step:initialize
    for current_num in numbers:
        updated_count = frequency_map.get(current_num, 0) + 1  # @step:increment-count
        frequency_map[current_num] = updated_count  # @step:increment-count
        if updated_count > threshold:  # @step:key-found
            return current_num
    return -1  # @step:complete

File: src/algorithms/hash-maps/counting/majority-element/sources/MajorityElement.cpp
// Majority Element — find the element that appears more than n/2 times using a frequency map
#include <vector>
#include <unordered_map>

int majorityElement(const std::vector<int>& numbers) {
    std::unordered_map<int, int> frequencyMap; // @step:initialize
    int threshold = (int)numbers.size() / 2; // @step:initialize
    for (int currentNum : numbers) {
        int updatedCount = ++frequencyMap[currentNum]; // @step:increment-count
        if (updatedCount > threshold) {
            return currentNum; // @step:key-found
        }
    }
    return -1; // @step:complete
}
